### ISSUE TITLE
Don't swallow errors

### ISSUE THREAD

**Issue by benjamingr** [pburtchaell/redux-promise-middleware#59]
It would be great to actually notify users on exceptions here: https://github.com/pburtchaell/redux-promise-middleware/blob/master/src/index.js#L54 vs recovering from the error.

At least a hook that logs rejections cleanly for debugging. Currently this is basically a:

    try {
        return [action, false] 
    } catch (e) {
        return [action, true]
    }

It would be useful to notify on rejections - maybe via taking a callback, so users don't get exceptions that are more silent by default.

**Comment by tomatau** [comment]
The caught error should still be passed through the dispatch and attached to the payload of the rejected action.

Do you have some concrete example or repro steps for an error thats hard to debug?

You can still put a then block on your own promise object before handing it off to the middleware -- more verbose I know but when debugging it's a useful step.

I'd be happy to expose some callback if it solves a problem here.

**Comment by benjamingr** [comment]
Exposing some callback or having a config option to fire global rejection events when such an error occurs if it's not handled both sound reasonable

**Comment by tomatau** [comment]
Please show some repro steps where the error action that's already emitted doesn't cover this.

**Comment by tomatau** [comment]
Going to close this as rejected actions are already available.

**Comment by tybro0103** [comment]
I think the problem is that when you access the promise returned by dispatching:
```javascript
const prom = store.dispatch(fetchUser()).payload.promise;
```
it will always be resolved because the errors are being caught on the line mentioned in OP.

True, you could use `prom.then(...)` and check to see if it returned an error inside of the `then`, but I'd expect my catch handler to work here still.

**Comment by tomatau** [comment]
You can put your own catch handler when you make the promise the first time:

```js
fetchUser() {
 return {
  payload: { promise: makePromise().then(_, catchHandler) }
 }
}
```

**Comment by tybro0103** [comment]
Unless I'm missing something, I don't think this will help me at the point where I'm doing the dispatch, will it?

Sure I can run some code in that catch handler, but it wont be in the same closure as where I'm calling `dispatch`.

**Comment by tybro0103** [comment]
Is there a reason to not return a rejected promise from dispatch?

**Comment by tomatau** [comment]
Yes, we are catching them so that we can handle failed actions without expecting people to always catch errors.

Why do you need to handle the error in the same closure as where the dispatch is being called?

**Comment by tybro0103** [comment]
Ah yes, I see. Brain fart.

I could see there being plenty of scenarios, but my particular one at the moment... I'm doing this in a router, and if the promise fails due to a 401 unauthorized, I want to redirect.

**Comment by tomatau** [comment]
You can redirect from the action creator using a history instance... or even have your action creator accept a callback for errors

**Comment by tybro0103** [comment]
I think I'll make do with checking within `then` for the error.

History instance could work, except my app is universal. I have my own ideas for handling routing universally: [Pouter](https://github.com/tybro0103/pouter)

**Comment by tomatau** [comment]
Universal apps can work fine with the history api - I know as I'm doing this already. :)

**Comment by tybro0103** [comment]
Oh, I just remembered what I was originally thinking before I got confused...

Why not just return the original promise instead of overwriting it? The original promise could have multiple `catch` attached all of which will be invoked upon a rejection so long as they aren't chained after a `then`.

This is contrived, but something like:

```
function doSomethingWithPromiseResolution(promise) {
  promise
    .then(/* dispatch success */)
    .catch(/* dispatch failure */);
  return promise; // additional catch() can be chained because original promise is returned
};
```

**Comment by benjamingr** [comment]
This was more about how if you don't explicitly catch the error in the reducer then you get a silenced error. By leaving the promise rejected promise libraries will report the error to the console. 

By forgetting a single error check (or having a single typo) you can get sucked into an hour long debugging session. 

The solution @tybro0103 would probably not trigger those default handlers so I do not think it solves this very well but he has the right idea.

**Comment by tybro0103** [comment]
@benjamingr oooo yes, I see what you're saying now. That is an even bigger issue! I think my problem is mostly the same because I'm actually looking to "rescue" from a particular error, but this middleware is already catching everything.

**Comment by benjamingr** [comment]
@tybro0103 I think the middleware catching everything makes sense given the point is to delegate to the reducer - this is the redux model.

The problem is that it makes errors silenced by default unless specifically handled where it took us a lot of time to make promises more debuggable by making errors report by default and adding rejection hooks.

Optimally, there would be a way to say "this event was fired and not handled by the reducer so report it" but I think that's impossible from the redux point of view unless we assume everyone using redux is using immutable state which would let us to query `state === state` in order to report it. It would be a big opinionated assumption on `redux-promise-middleware` and one that @pburtchaell would have to either make or not. 

I just think the current solution isn't too great for having code that's easy to debug. I opened this issue mainly because someone at my company used this middleware and I only knew how to find the issue since I have promise library experience.

**Comment by tybro0103** [comment]
@benjamingr I agree. 

I know this library has to catch in order for it to be able to dispatch the failure action. I think there's two problems with that:

1. The errors being silenced. They would otherwise be unhandled rejections, but since this library catches them they're considered handled.
2. The inability to have this library catch the error AND your own code catch the error.

Not sure what the solution is. Perhaps this library could return a new rejected promise?

**Comment by tybro0103** [comment]
Hmmm, I guess you were saying something along the lines of having the error put into the redux state would be handling it?

**Comment by benjamingr** [comment]
Yes, exactly - it is not sileencing rejections it's moving them to the reducer. The problem here is only with the _default_ being converted from "report unhandled errors" to "silence unhandled errors".

**Comment by tomatau** [comment]
FWIW I strongly recommend you log you actions for debugging. An action is always thrown for any rejected promise.

**Comment by benjamingr** [comment]
Right, but logging all actions is problematic if you have a large app with a lot of actions - it's easy to get to a point it's a ton of noise.

**Comment by tomatau** [comment]
Most loggers provide filters for noise actions like textfield updates. Allowing you to log the async stuff.

**Comment by pburtchaell** [comment]
So I have not had a chance to read through this entire conversation, but I am working on a version 3.0.0 release of the middleware. The proposed changes will enable both fulfilled and rejected handlers.

For example:

```js
// fulfilled promise
const foo = () => {
  return dispatch => {

    return dispatch({
      type: 'FOO',
      payload: new Promise(resolve => {
        resolve('foo'); // resolve the promise with the value 'foo'
      })
    }).then(({ value, action }) => {
      console.log(value); // => 'foo'
      console.log(action.type); // => 'FOO_FULFILLED'
    });
  };
}

// rejected promise
const bar = () => {
  return dispatch => {

    return dispatch({
      type: 'BAR',
      payload: new Promise(reject => {
        reject('bar'); // reject the promise for the reason 'bar'
      })
    }).then(() => null, ({ reason, action }) => {
      console.log(reason); // => 'bar'
      console.log(action.type); // => 'BAR_REJECTED'
      console.log(action.error); // => true
    });
  };
}
```

I have not implemented a way to use `.catch()` yet, but I think having a rejected handler will be sufficient.

**Comment by benjamingr** [comment]
Great.

**Comment by pburtchaell** [comment]
This API is released as `redux-promise-middleware@3.0.0`. Please let me know if you have any issues.

**Comment by zowers** [comment]
the middleware does not throw new error with `reason` and `action` anymore, it rethrows original error, so handling code can only use `.catch(err => {})` since [4.0.0](https://github.com/pburtchaell/redux-promise-middleware/releases/tag/4.0.0) after #105
